From 8800ecdfd83cd3e39e5d676186e7823459059236 Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Tue, 26 Jan 2010 17:48:22 -0800
Subject: [PATCH 2/3] fix leap second message hang

The system will hang in case of leap second insertion when
CONFIG_PMEM_LOG_REG is enabled.

   ntp_leap_second
     write_seqlock_irq(&xtime_lock) <-- grab it
       printk
         release_console_sem
           call_console_drivers
             pmem_write_data
               pmem_write_log_desc
                 do_gettimeofday
                   getnstimeofday
                     read_seqbegin(&xtime_lock) <-- deadlock

This patch moves printk outside of holding xtime_lock to prevent from
deadlock.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 kernel/time/ntp.c |   11 +++++++----
 1 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/kernel/time/ntp.c b/kernel/time/ntp.c
index 7c0f180..a83ecff 100644
--- a/kernel/time/ntp.c
+++ b/kernel/time/ntp.c
@@ -187,6 +187,7 @@ void ntp_clear(void)
 static enum hrtimer_restart ntp_leap_second(struct hrtimer *timer)
 {
 	enum hrtimer_restart res = HRTIMER_NORESTART;
+	int leap = 0;
 
 	write_seqlock(&xtime_lock);
 
@@ -196,8 +197,7 @@ static enum hrtimer_restart ntp_leap_second(struct hrtimer *timer)
 	case TIME_INS:
 		timekeeping_leap_insert(-1);
 		time_state = TIME_OOP;
-		printk(KERN_NOTICE
-			"Clock: inserting leap second 23:59:60 UTC\n");
+		leap = 1;
 		hrtimer_add_expires_ns(&leap_timer, NSEC_PER_SEC);
 		res = HRTIMER_RESTART;
 		break;
@@ -205,8 +205,7 @@ static enum hrtimer_restart ntp_leap_second(struct hrtimer *timer)
 		timekeeping_leap_insert(1);
 		time_tai--;
 		time_state = TIME_WAIT;
-		printk(KERN_NOTICE
-			"Clock: deleting leap second 23:59:59 UTC\n");
+		leap = -1;
 		break;
 	case TIME_OOP:
 		time_tai++;
@@ -220,6 +219,10 @@ static enum hrtimer_restart ntp_leap_second(struct hrtimer *timer)
 
 	write_sequnlock(&xtime_lock);
 
+	if (leap != 0)
+		printk(KERN_NOTICE "Clock: %sing leap second 23:59:%d UTC\n",
+			leap == 1 ? "insert" : "delet", leap == 1 ? 60 : 59);
+
 	return res;
 }
 
-- 
1.6.5.2

